### PR TITLE
bpo-34663: Add support for POSIX_SPAWN_USEVFORK in posix_spawn when available

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue34663

### CODE REVIEW

File: Modules/posixmodule.c [hunk 1]
```diff
@@ -5278,6 +5278,16 @@ parse_posix_spawn_flags(PyObject *setpgroup, int resetids, PyObject *setsigmask,
 #endif
     }
 
+    if (usevfork) {
+#ifdef POSIX_SPAWN_USEVFORK
+        all_flags |= POSIX_SPAWN_SETSCHEDULER;
+#else
+        PyErr_SetString(PyExc_NotImplementedError,
+                "The UseVFork option is not supported in this system.");
```
Comment (vstinner) [inline Modules/posixmodule.c]: The parameter is called "usevfork", the error message should use the same name (not "UseVFork").

File: Modules/posixmodule.c [hunk 2]
```diff
@@ -5425,6 +5435,8 @@ os.posix_spawn
         The sigmask to use with the POSIX_SPAWN_SETSIGDEF flag.
     scheduler: object = NULL
         A tuple with the scheduler policy (optional) and parameters.
+    usevfork: bool(accept={int}) = False
```
Comment (vstinner) [inline Modules/posixmodule.c]: I suggest the name "use_vfork", or maybe even "vfork".

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by pablogsal** [comment]
I have made the requested changes; please review again

File: Modules/posixmodule.c [hunk 3]
```diff
@@ -5278,6 +5278,16 @@ parse_posix_spawn_flags(PyObject *setpgroup, int resetids, PyObject *setsigmask,
 #endif
     }
 
+    if (use_vfork) {
+#ifdef POSIX_SPAWN_USEVFORK
+        all_flags |= POSIX_SPAWN_SETSCHEDULER;
+#else
+        PyErr_SetString(PyExc_NotImplementedError,
```
Comment (vstinner) [inline Modules/posixmodule.c:5285]: You can use argument_unavailable_error() here.

File: Modules/posixmodule.c [hunk 4]
```diff
@@ -5425,6 +5435,8 @@ os.posix_spawn
         The sigmask to use with the POSIX_SPAWN_SETSIGDEF flag.
     scheduler: object = NULL
         A tuple with the scheduler policy (optional) and parameters.
+    use_vfork: bool(accept={int}) = False
+        If the value is `True` the POSIX_SPAWN_USEVFORK will be activated.
```
Comment (vstinner) [inline Modules/posixmodule.c]: "If the value is true": use_vfork=1 should be equivalent to use_vfork=True.

File: Lib/test/test_posix.py [hunk 5]
```diff
@@ -1790,6 +1790,19 @@ def test_dup2(self):
         with open(dupfile) as f:
             self.assertEqual(f.read(), 'hello')
 
+    def test_vfork(self):
+        try:
+            pid = posix.posix_spawn(
```
Comment (vstinner) [inline Lib/test/test_posix.py]: You might use args = self.python_args() and  posix.posix_spawn(args[0], args, ...)

**Comment by vstinner** [comment]
Please also document the use_vfork parameter in Doc/library/os.html

File: Doc/library/os.rst [hunk 6]
```diff
@@ -3465,6 +3465,10 @@ written in Python, such as a mail server's external command delivery program.
    :c:data:`POSIX_SPAWN_SETSCHEDPARAM` and :c:data:`POSIX_SPAWN_SETSCHEDULER`
    flags.
 
+  If the value of *use_vfork* is true, the child will be created using
+  :c:func:`vfork` instead of :c:func:`fork`. This corresponds to the
+  GNU specific flag :c:data:`POSIX_SPAWN_USEVFORK`.
+
```
Comment (vstinner) [inline Doc/library/os.rst]: Oh, "Availability: ..." is missing: this function is not available on Windows, right?

I would suggest to document in "Availability: ..." that the flag is only available on Linux.

File: Doc/library/os.rst [hunk 7]
```diff
@@ -3465,6 +3465,10 @@ written in Python, such as a mail server's external command delivery program.
    :c:data:`POSIX_SPAWN_SETSCHEDPARAM` and :c:data:`POSIX_SPAWN_SETSCHEDULER`
    flags.
 
+  If the value of *use_vfork* is true, the child will be created using
+  :c:func:`vfork` instead of :c:func:`fork`. This corresponds to the
+  GNU specific flag :c:data:`POSIX_SPAWN_USEVFORK`.
```
Comment (vstinner) [inline Doc/library/os.rst]: You should document that the function raises an exception if the flag is unavailable on the platform.

File: Lib/test/test_posix.py [hunk 8]
```diff
@@ -1790,6 +1790,20 @@ def test_dup2(self):
...
+        args = self.python_args()
+        try:
+            pid = posix.posix_spawn(
+                args[0],
+                args,
+                os.environ,
+                use_vfork=True
+            )
+        except NotImplementedError:
+            unittest.SkipTest("POSIX_SPAWN_USEVFORK not available on this system")
```
Comment (vstinner) [inline Lib/test/test_posix.py]: Hum, you can just reuse the exception message as part of this error message.

File: Misc/NEWS.d/next/Core and Builtins/2018-09-13-11-14-38.bpo-34663.BtJz-5.rst [hunk 9]
```diff
@@ -0,0 +1,2 @@
+Added support for `usevfork` parameters of `posix_spawn`, which allows to
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2018-09-13-11-14-38.bpo-34663.BtJz-5.rst]: use_vfork with underscore, no?

`posix_spawn` can be written :func:`os.posix_spawn` to get a link to the function in the generated doc.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Doc/library/os.rst [hunk 10]
```diff
@@ -3467,7 +3467,10 @@ written in Python, such as a mail server's external command delivery program.
 
   If the value of *use_vfork* is true, the child will be created using
   :c:func:`vfork` instead of :c:func:`fork`. This corresponds to the
-  GNU specific flag :c:data:`POSIX_SPAWN_USEVFORK`.
+  GNU specific flag :c:data:`POSIX_SPAWN_USEVFORK`. If the flag is not
+  available on the platform, :exc:`NotImplementedError` will be raised.
+
+   Availability: Using *use_vfork* is only available on Linux.
```
Comment (vstinner) [inline Doc/library/os.rst]: You should document that the function is only available on Unix. Example:
https://docs.python.org/dev/library/os.html#os.fork

File: Lib/test/test_posix.py [hunk 11]
```diff
@@ -1791,16 +1791,16 @@ def test_dup2(self):
...
         try:
             pid = posix.posix_spawn(
                 args[0],
                 args,
                 os.environ,
                 use_vfork=True
             )
-        except NotImplementedError:
-            unittest.SkipTest("POSIX_SPAWN_USEVFORK not available on this system")
+        except NotImplementedError as e:
```
Comment (vstinner) [inline Lib/test/test_posix.py]: nitpick: i prefer to avoid single character variable, i suggest the name "exc".

File: Lib/test/test_posix.py [hunk 12]
```diff
@@ -1791,16 +1791,16 @@ def test_dup2(self):
...
+            raise unittest.SkipTest(e)
```
Comment (vstinner) [inline Lib/test/test_posix.py]: You can use: raise self.skipTest(str(exc)).

File: Modules/posixmodule.c (hunk 3 above)
Comment (vstinner) [reply Modules/posixmodule.c:5285]: I discussed with Pablo and using "posix_spawn()" function name is less useless than mentioning the "POSIX_SPAWN_USEVFORK flag".

**Review by vstinner — APPROVED** [review]
LGTM.

@gpshead: Would you mind to double check this PR?

@benjaminp: Would you be interested to have a look as well? (the plan is then to experiment to use it in subprocess for performance.)

**Pull request closed without merging**